Port c++ tools to python in cli_tools_py

mikaelarguedas opened the pull request:
depends on https://github.com/ros2/rclpy/pull/71

Karsten1987 commented:
I like that. 
Should this PR also remove the existing c++ implementation to avoid duplication?

Karsten1987 commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,27 @@
+    rclpy.init([])
+
+    node = rclpy.create_node('rosnode_list')
+    time.sleep(args.time)
```
is that `time.sleep` the exactly equivalent of `rclcpp::spin()`? Does the DDS impl actually turn with that?

wjwwood replied:
No, `time.sleep` is part of Python's builtin `time` module. So no to both questions.

dirk-thomas commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,27 @@
+                print("Name: %s" % node)
+            print('Number of currently available nodes: %d' % len(a))
+        else:
+            print("%s" % a)
```
Isn't this the same as `print(a)`? Would printing each node name on a separate line be better?

dirk-thomas commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,27 @@
+
+    node = rclpy.create_node('rosnode_list')
+    time.sleep(args.time)
+    if(rclpy.ok()):
```
No parenthesis around the condition.

dirk-thomas commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,27 @@
+    parser.add_argument('-t', '--time', default=1,
+                        help='spin time to collect nodes')
+    parser.add_argument('-v', '--verbose', action='store_true',
+                        help='verbos output print')
```
Spelling. I more specific help text what extra output to expect would be nice.

dirk-thomas commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,27 @@
+    node = rclpy.create_node('rosnode_list')
+    time.sleep(args.time)
+    if(rclpy.ok()):
+        a = node.get_node_names()
```
Better variable name than `a`, probably `node_names`.

dirk-thomas commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,27 @@
+        if args.verbose:
+            for node in a:
+                print("Name: %s" % node)
+            print('Number of currently available nodes: %d' % len(a))
```
Is the verbose output really helpful? If the node names are printed on separate lines ` | wc -l` achieve the same thing.

mikaelarguedas replied:
@dirk-thomas I don't have strong feelings about this, I was mimicking the behviour of the existing c++ tool but am totally open to suggestions for improvements

mikaelarguedas commented:
Sorry all, that was a direct copy paste from an old script I had lying around from 6 months ago.
I'll review it myself, address all your comments and put it back in review after that.

mikaelarguedas commented:
I updated the PR based on feedback.
Regarding https://github.com/ros2/cli_tools/pull/9#discussion_r108307667 I just use the timeout exposed by `rclpy.spin_once` but I can implement a timer if you guys think it's the thing to do. It goes back to "do we want to expose timeouts in spin functions" (related to https://github.com/ros2/rclcpp/issues/280#issuecomment-289627771)

dirk-thomas commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,37 @@
+    rclpy.spin_once(node, timeout_sec=args.time)
+    if rclpy.ok():
+        node_names = node.get_node_names()
+        print(*node_names, sep='\n')
```
Even more elegant :1st_place_medal:

mikaelarguedas commented:
ready for review

wjwwood commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,37 @@
+    rclpy.init([])
+
+    node = rclpy.create_node('rosnode_list')
+    rclpy.spin_once(node, timeout_sec=args.time)
```
What if there is something to do and it finishes doing it almost immediately? This script won't end up waiting `args.time` in that case.

mikaelarguedas replied:
the waitset being empty except for the SIGINT guard condition I don't think it can exit before `args.time` has been waited for. but as I said in https://github.com/ros2/cli_tools/pull/9#issuecomment-289851561 I don't mind throwing in a timer here and use the callback to trigger the "get_topic_names_and_types"

dirk-thomas replied:
I think the point is that the function `spin_once` doesn't guarantee that and in the future there could be other things in the waitset by default. So a timer would be good I think.

wjwwood commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,37 @@
+                        help='display the number of nodes discovered')
+    args = parser.parse_args()
+
+    rclpy.init([])
```
I know it doesn't do anything now, buy why not pass `sys.argv` to this?

mikaelarguedas replied:
I think this is equivalent even in the scenario we decide to do something with the command line arguments: https://github.com/ros2/rclpy/blob/8aabbd2037a88034a2f0b85e3fe751f4515608a3/rclpy/rclpy/__init__.py#L35

wjwwood commented on cli_tools_py/rostopic_list.py:
```diff
@@ -0,0 +1,37 @@
+                        help='display the number of topics discovered')
+    args = parser.parse_args()
+
+    rclpy.init([])
```
Same.

mikaelarguedas replied:
Same https://github.com/ros2/cli_tools/pull/9#discussion_r108712113

wjwwood commented on cli_tools_py/rostopic_list.py:
```diff
@@ -0,0 +1,37 @@
+    rclpy.init([])
+
+    node = rclpy.create_node('rostopic_list')
+    rclpy.spin_once(node, timeout_sec=args.time)
```
Same.

mikaelarguedas replied:
same https://github.com/ros2/cli_tools/pull/9#discussion_r108712113

dirk-thomas commented on cli_tools_py/rostopic_list.py:
```diff
@@ -0,0 +1,37 @@
+    rclpy.spin_once(node, timeout_sec=args.time)
+    if rclpy.ok():
+        topic_names = node.get_topic_names_and_types()
+        print(topic_names)
```
Should the topic names be printed on separate lines as for `rosnode_list`?

mikaelarguedas replied:
I guess I should call this variable "topic_names_and_types". This will call __str__ on a TopicNamesAndTypes object defined [here](https://github.com/ros2/rclpy/pull/71/files#diff-52d9a2b366601e50c4c1bb509abb5b44R32).
So topic_names will be printed on separate lines

dirk-thomas replied:
Oh, I see. I didn't get it when reading the variable name. And didn't read the full line above either :wink: Renaming the variable sounds like a great idea.

mikaelarguedas commented:
I addressed review comments. One last thing I'd like to have your opinion on: should rosnode list filter out it's own node for the list ?

Right now if you run talker / listener example you get:
```
rosnode_list
listener
talker
Number of currently available nodes: 3
```

dirk-thomas commented:
> should rosnode list filter out it's own node for the list ?

Imo: yes.

dirk-thomas commented on cli_tools_py/rosnode_list.py:
```diff
@@ -0,0 +1,50 @@
+        nonlocal timeout_reached
+        timeout_reached = True
+
+    rclpy.init(sys.argv)
```
I am not sure if passing `sys.argv` explicitly should be done. The API defaults to that value already: https://github.com/ros2/rclpy/blob/8d11493b60d8c02145b6dda29457ca9eb6f1089f/rclpy/rclpy/__init__.py#L35

mikaelarguedas replied:
I agree but we seem to have different opinions here: https://github.com/ros2/cli_tools/pull/9#discussion_r108712113

dirk-thomas replied:
Well, the referenced comment from @wjwwood suggested *not* to pass `[]` which I agree with. I think passing `sys.argv` explicitly vs. using the default value of the argument is a bit different.

mikaelarguedas replied:
I'm agnostic to either expose it all the way or not exposing them at all, but am not satisfied with my current implementation: either we pass them as default argument to the main function and pass them along to the `rclpy.init` call. or we rely on the underlying behavior in `rclpy.init()` to take them if nothing else is provided. I'm happy to make the change one way or the other

mikaelarguedas replied:
oh I misinterpreted @wjwwood comment then :S my bad. Agreed that `[]` is bad and should be removed here and every legacy examples.

wjwwood replied:
Personally, I like the explicit passing of `sys.argv` always, because otherwise you have no idea that `sys.argv` is being considered (and possibly mutated) by `rclpy.init()`. However, this is a common occurrence in Python, like [`argparse.parse_args()`](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.parse_args).

Ultimately, I think it is wise to have it default to `sys.argv` implicitly is the right thing to do, for convenience and because that's how we did it in ROS 1. I'd support passing nothing here as well as in the other place I originally commented.

wjwwood commented:
> should rosnode list filter out it's own node for the list ?

You could also extend this question to be "should it show other `rosnode_list` which are running at the same time?" or "should it show a `rostopic_list` or `rostopic_echo` that is running at the same time?". I know those aren't the same thing, but this is the idea behind this section about "hidden" topics:

http://design.ros2.org/articles/topic_and_service_names.html#hidden-topic-or-service-names

You could imagine a similar rule for node names, like "if a node name starts with an `_`, then it is hidden by most tools by default". Then `rosnode_list` could have an option to "show all" despite this, in which case I would say it should show itself in the list.

At the heart of the question is: if you hide the `rosnode_list` node from the output do you do so because all `rosnode_list` nodes should be hidden (because the fact that they use a node is an implementation detail?) or because it should self filter itself?

I'm not sure I agree with the self filtering. That could be the default behavior for convenience, but in that case I would argue it should have an option to not self filter. And generally I feel that having a convention for "hiding" certain nodes from that list and then reusing that convention to implement the self filtering would be better.

mikaelarguedas commented:
That's a good point. I like the idea of the tools supporting hidden nodes and hidden topics. I added the filtering capabitlity to both `rostopic_list` and `rosnode_list` with a todo to revisit them once we have a design specification for hidden nodes

wjwwood commented:
> That's a good point. I like the idea of the tools supporting hidden nodes and hidden topics. I added the filtering capabitlity to both rostopic_list and rosnode_list with a todo to revisit them once we have a design specification for hidden nodes.

Sounds good to me, my +1 review status stands.

mikaelarguedas commented:
Thanks, merging then
